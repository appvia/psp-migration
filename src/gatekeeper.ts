import * as k8s from '@kubernetes/client-node'

import * as mod from './gatekeeper'

export function transform_gatekeeper(PSP: k8s.V1beta1PodSecurityPolicy): object[] {
  const policies = []
  if (PSP.spec?.allowPrivilegeEscalation === false)
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPAllowPrivilegeEscalationContainer'))

  if (PSP.spec?.privileged === false)
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPPrivilegedContainer'))

  if (PSP.spec?.readOnlyRootFilesystem === true)
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPReadOnlyRootFilesystem'))

  if (PSP.spec?.hostIPC === false || PSP.spec?.hostPID === false)
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPHostNamespace'))

  if (PSP.spec?.hostPorts)
    PSP.spec?.hostPorts.forEach(portRange =>
      policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPHostNetworkingPorts', { hostNetwork: true, min: portRange.min, max: portRange.max }))
    )

  if (PSP.spec?.hostNetwork !== undefined)
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPHostNetworkingPorts', { hostNetwork: PSP.spec.hostNetwork }))

  if (!PSP.spec?.volumes?.includes('*'))
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPVolumeTypes', { volumes: PSP.spec?.volumes }))

  if (PSP.metadata?.annotations && PSP.metadata?.annotations['apparmor.security.beta.kubernetes.io/allowedProfileNames'])
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPAppArmor', { allowedProfiles: PSP.metadata?.annotations['apparmor.security.beta.kubernetes.io/allowedProfileNames'].split(',') }))

  if (PSP.metadata?.annotations && PSP.metadata?.annotations['seccomp.security.alpha.kubernetes.io/allowedProfileNames'])
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPSeccomp', { allowedProfiles: PSP.metadata?.annotations['seccomp.security.alpha.kubernetes.io/allowedProfileNames'].split(',') }))

  if (PSP.spec?.seLinux?.rule === 'MustRunAs')
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPSELinuxV2', { allowedSELinuxOptions: [PSP.spec?.seLinux?.seLinuxOptions] }))

  if (PSP.spec?.allowedCapabilities || PSP.spec?.requiredDropCapabilities)
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPCapabilities', { allowedCapabilities: (PSP.spec?.allowedCapabilities || []), requiredDropCapabilities: (PSP.spec?.requiredDropCapabilities || []) }))

  if (PSP.spec?.allowedFlexVolumes)
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPFlexVolumes', { allowedFlexVolumes: PSP.spec?.allowedFlexVolumes }))

  if (PSP.spec?.allowedHostPaths)
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPHostFilesystem', { allowedHostPaths: PSP.spec?.allowedHostPaths }))

  if (PSP.spec?.allowedProcMountTypes)
    PSP.spec?.allowedProcMountTypes.forEach(procMountType =>
      policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPProcMount', { procMount: procMountType })))

  if (PSP.spec?.allowedUnsafeSysctls || PSP.spec?.forbiddenSysctls)
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPForbiddenSysctls', { allowedSysctls: (PSP.spec?.allowedUnsafeSysctls || []), forbiddenSysctls: (PSP.spec?.forbiddenSysctls || []) }))

  if (PSP.spec?.runAsUser && PSP.spec?.runAsUser?.rule !== 'RunAsAny')
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPAllowedUsers', { runAsUser: PSP.spec?.runAsUser }))

  if (PSP.spec?.runAsGroup && PSP.spec?.runAsGroup?.rule !== 'RunAsAny')
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPAllowedUsers', { runAsGroup: PSP.spec?.runAsGroup }))

  if (PSP.spec?.supplementalGroups && PSP.spec?.supplementalGroups?.rule !== 'RunAsAny')
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPAllowedUsers', { supplementalGroups: PSP.spec?.supplementalGroups }))

  if (PSP.spec?.fsGroup && PSP.spec?.fsGroup?.rule !== 'RunAsAny')
    policies.push(mod.gatekeeper_pod_policy_helper('K8sPSPAllowedUsers', { fsGroup: PSP.spec?.fsGroup }))

  if (PSP.spec?.defaultAddCapabilities)
    policies.push({
      apiVersion: "mutations.gatekeeper.sh/v1beta1",
      kind: "ModifySet",
      metadata: { name: "psp-k8spspdefaultaddcapabilities" },
      spec: {
        applyTo: [{ groups: [""], versions: ["v1"], kinds: ["Pod"] }],
        match: { scope: "Namespaced", kinds: [{ apiGroups: ["*"], kinds: ["Pod"] }] },
        location: "spec.containers[name:*].securityContext.capabilities.add",
        parameters: {
          values: {
            fromList: PSP.spec?.defaultAddCapabilities || []
          }
        }
      }
    })
  if (PSP.spec?.defaultAllowPrivilegeEscalation !== undefined) {
    policies.push({
      apiVersion: "mutations.gatekeeper.sh/v1beta1",
      kind: "ModifySet",
      metadata: { name: "psp-k8spspdefaultallowprivilegeescalation" },
      spec: {
        applyTo: [{ groups: [""], versions: ["v1"], kinds: ["Pod"] }],
        location: "spec.containers[name:*].securityContext.allowPrivilegeEscalation",
        parameters: {
          pathTests: [{
            subPath: "spec.containers[name:*].securityContext.allowPrivilegeEscalation",
            condition: "MustNotExist"
          }],
          assign: { value: PSP.spec?.defaultAllowPrivilegeEscalation }
        }
      }
    })
    policies.push({
      apiVersion: "mutations.gatekeeper.sh/v1beta1",
      kind: "ModifySet",
      metadata: { name: "psp-k8spspdefaultallowprivilegeescalation-init" },
      spec: {
        applyTo: [{ groups: [""], versions: ["v1"], kinds: ["Pod"] }],
        location: "spec.initContainers[name:*].securityContext.allowPrivilegeEscalation",
        parameters: {
          pathTests: [{
            subPath: "spec.initContainers[name:*].securityContext.allowPrivilegeEscalation",
            condition: "MustNotExist"
          }],
          assign: { value: PSP.spec?.defaultAllowPrivilegeEscalation }
        }
      }
    })
  }
  return policies
}


export function gatekeeper_pod_policy_helper(kind: string, parameters: object | null = null): object {
  return {
    apiVersion: "constraints.gatekeeper.sh/v1beta1",
    kind: kind,
    metadata: {
      name: `psp-${kind.toLowerCase()}`,
    },
    spec: {
      match: {
        kinds: [
          {
            apiGroups: [""],
            kinds: ["Pod"]
          }
        ]
      },
      parameters: parameters
    }
  }
}
