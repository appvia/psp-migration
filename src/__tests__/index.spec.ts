import * as mod from '../index'
import * as gatekeeper from '../gatekeeper'
import * as fs from 'fs'

const fixturePSPYAML = `
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: policy
spec:
  runAsUser:
    rule: RunAsAny
  seLinux:
    rule: RunAsAny
  fsGroup:
    rule: RunAsAny
  supplementalGroups:
    rule: 'RunAsAny'
  volumes:
    - '*'
`
const fixturePSPJSON = `{ "apiVersion": "policy/v1beta1", "kind": "PodSecurityPolicy", "metadata": { "name": "policy" }, "spec": { "runAsUser": { "rule": "RunAsAny" }, "seLinux": { "rule": "RunAsAny" }, "fsGroup": { "rule": "RunAsAny" }, "supplementalGroups": { "rule": "RunAsAny" }, "volumes": ["*"] } }`
const fixturePSPObject = { "apiVersion": "policy/v1beta1", "kind": "PodSecurityPolicy", "metadata": { "name": "policy" }, "spec": { "runAsUser": { "rule": "RunAsAny" }, "seLinux": { "rule": "RunAsAny" }, "fsGroup": { "rule": "RunAsAny" }, "supplementalGroups": { "rule": "RunAsAny" }, "volumes": ["*"] } }

describe('parse', () => {

  it('should parse a yaml object', () => expect(mod.parse(fixturePSPYAML)).toMatchSnapshot())

  it('should parse a json object', () => expect(mod.parse(fixturePSPJSON)).toMatchSnapshot())
})

describe('transform', () => {
  it('should call the right engine', () => {
    const spy = jest.spyOn(gatekeeper, "transform_gatekeeper")
    mod.transform(fixturePSPObject, 'gatekeeper')
    expect(spy).toHaveBeenCalled()
  })
})

function help_load_psp(fixture: string) {
  const yaml = fs.readFileSync(`tests/${fixture}/psp.yaml`, { flag: 'r' })
  return mod.parse(yaml.toString())
}

const pspFields = [
  "allowPrivilegeEscalation",
  "allowedCapabilities",
  "allowedFlexVolumes",
  "allowedHostPaths",
  "allowedProcMountTypes",
  "allowedUnsafeSysctls",
  "apparmor",
  "defaultAddCapabilities",
  "defaultAllowPrivilegeEscalation",
  "forbiddenSysctls",
  "fsgroup",
  "hostIPC",
  "hostNetwork",
  "hostPID",
  "hostPorts",
  "privileged",
  "readOnlyRootFilesystem",
  "requiredDropCapabilities",
  "runAsGroup",
  "runAsUser",
  "seLinux",
  "seccomp",
  "supplementalGroups",
  "volumes"
].map(field => [field])

describe('transform_gatekeeper', () => {
  it('should do an empty PSP', () => expect(gatekeeper.transform_gatekeeper(fixturePSPObject)).toStrictEqual([]))
  test.each(pspFields)('%s', (field) =>
    expect(gatekeeper.transform_gatekeeper(help_load_psp(field))).toMatchSnapshot()
  )
})