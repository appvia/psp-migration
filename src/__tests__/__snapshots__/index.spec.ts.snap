// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should parse a json object 1`] = `
Object {
  "apiVersion": "policy/v1beta1",
  "kind": "PodSecurityPolicy",
  "metadata": Object {
    "name": "policy",
  },
  "spec": Object {
    "fsGroup": Object {
      "rule": "RunAsAny",
    },
    "runAsUser": Object {
      "rule": "RunAsAny",
    },
    "seLinux": Object {
      "rule": "RunAsAny",
    },
    "supplementalGroups": Object {
      "rule": "RunAsAny",
    },
    "volumes": Array [
      "*",
    ],
  },
}
`;

exports[`parse should parse a yaml object 1`] = `
Object {
  "apiVersion": "policy/v1beta1",
  "kind": "PodSecurityPolicy",
  "metadata": Object {
    "name": "policy",
  },
  "spec": Object {
    "fsGroup": Object {
      "rule": "RunAsAny",
    },
    "runAsUser": Object {
      "rule": "RunAsAny",
    },
    "seLinux": Object {
      "rule": "RunAsAny",
    },
    "supplementalGroups": Object {
      "rule": "RunAsAny",
    },
    "volumes": Array [
      "*",
    ],
  },
}
`;

exports[`transform_gatekeeper allowPrivilegeEscalation 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPAllowPrivilegeEscalationContainer",
    "metadata": Object {
      "name": "psp-k8spspallowprivilegeescalationcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
]
`;

exports[`transform_gatekeeper allowedCapabilities 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPCapabilities",
    "metadata": Object {
      "name": "psp-k8spspcapabilities",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedCapabilities": Array [
          "something",
        ],
        "requiredDropCapabilities": Array [],
      },
    },
  },
]
`;

exports[`transform_gatekeeper allowedFlexVolumes 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPFlexVolumes",
    "metadata": Object {
      "name": "psp-k8spspflexvolumes",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedFlexVolumes": Array [
          Object {
            "driver": "example/lvm",
          },
        ],
      },
    },
  },
]
`;

exports[`transform_gatekeeper allowedHostPaths 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPPrivilegedContainer",
    "metadata": Object {
      "name": "psp-k8spspprivilegedcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPHostFilesystem",
    "metadata": Object {
      "name": "psp-k8spsphostfilesystem",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedHostPaths": Array [
          Object {
            "pathPrefix": "/foo",
            "readOnly": true,
          },
        ],
      },
    },
  },
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPHostFilesystem",
    "metadata": Object {
      "name": "psp-k8spsphostfilesystem",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedHostPaths": Array [
          Object {
            "pathPrefix": "/foo",
            "readOnly": true,
          },
        ],
      },
    },
  },
]
`;

exports[`transform_gatekeeper allowedProcMountTypes 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPPrivilegedContainer",
    "metadata": Object {
      "name": "psp-k8spspprivilegedcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPProcMount",
    "metadata": Object {
      "name": "psp-k8spspprocmount",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "procMount": "Default",
      },
    },
  },
]
`;

exports[`transform_gatekeeper allowedUnsafeSysctls 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPForbiddenSysctls",
    "metadata": Object {
      "name": "psp-k8spspforbiddensysctls",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedSysctls": Array [
          "kernel.msgmax",
        ],
        "forbiddenSysctls": Array [],
      },
    },
  },
]
`;

exports[`transform_gatekeeper apparmor 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPAppArmor",
    "metadata": Object {
      "name": "psp-k8spspapparmor",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedProfiles": Array [
          "runtime/default",
        ],
      },
    },
  },
]
`;

exports[`transform_gatekeeper defaultAddCapabilities 1`] = `
Array [
  Object {
    "apiVersion": "mutations.gatekeeper.sh/v1beta1",
    "kind": "ModifySet",
    "metadata": Object {
      "name": "psp-k8spspdefaultaddcapabilities",
    },
    "spec": Object {
      "applyTo": Array [
        Object {
          "groups": Array [
            "",
          ],
          "kinds": Array [
            "Pod",
          ],
          "versions": Array [
            "v1",
          ],
        },
      ],
      "location": "spec.containers[name:*].securityContext.capabilities.add",
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "*",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
        "scope": "Namespaced",
      },
      "parameters": Object {
        "values": Object {
          "fromList": Array [
            "something",
          ],
        },
      },
    },
  },
]
`;

exports[`transform_gatekeeper defaultAllowPrivilegeEscalation 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPAllowPrivilegeEscalationContainer",
    "metadata": Object {
      "name": "psp-k8spspallowprivilegeescalationcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
  Object {
    "apiVersion": "mutations.gatekeeper.sh/v1beta1",
    "kind": "ModifySet",
    "metadata": Object {
      "name": "psp-k8spspdefaultallowprivilegeescalation",
    },
    "spec": Object {
      "applyTo": Array [
        Object {
          "groups": Array [
            "",
          ],
          "kinds": Array [
            "Pod",
          ],
          "versions": Array [
            "v1",
          ],
        },
      ],
      "location": "spec.containers[name:*].securityContext.allowPrivilegeEscalation",
      "parameters": Object {
        "assign": Object {
          "value": false,
        },
        "pathTests": Array [
          Object {
            "condition": "MustNotExist",
            "subPath": "spec.containers[name:*].securityContext.allowPrivilegeEscalation",
          },
        ],
      },
    },
  },
  Object {
    "apiVersion": "mutations.gatekeeper.sh/v1beta1",
    "kind": "ModifySet",
    "metadata": Object {
      "name": "psp-k8spspdefaultallowprivilegeescalation-init",
    },
    "spec": Object {
      "applyTo": Array [
        Object {
          "groups": Array [
            "",
          ],
          "kinds": Array [
            "Pod",
          ],
          "versions": Array [
            "v1",
          ],
        },
      ],
      "location": "spec.initContainers[name:*].securityContext.allowPrivilegeEscalation",
      "parameters": Object {
        "assign": Object {
          "value": false,
        },
        "pathTests": Array [
          Object {
            "condition": "MustNotExist",
            "subPath": "spec.initContainers[name:*].securityContext.allowPrivilegeEscalation",
          },
        ],
      },
    },
  },
]
`;

exports[`transform_gatekeeper forbiddenSysctls 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPForbiddenSysctls",
    "metadata": Object {
      "name": "psp-k8spspforbiddensysctls",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedSysctls": Array [
          "foo.*",
        ],
        "forbiddenSysctls": Array [
          "kernel.*",
        ],
      },
    },
  },
]
`;

exports[`transform_gatekeeper fsgroup 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPPrivilegedContainer",
    "metadata": Object {
      "name": "psp-k8spspprivilegedcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPAllowedUsers",
    "metadata": Object {
      "name": "psp-k8spspallowedusers",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "fsGroup": Object {
          "ranges": Array [
            Object {
              "max": 200,
              "min": 100,
            },
          ],
          "rule": "MustRunAs",
        },
      },
    },
  },
]
`;

exports[`transform_gatekeeper hostIPC 1`] = `Array []`;

exports[`transform_gatekeeper hostNetwork 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPHostNetworkingPorts",
    "metadata": Object {
      "name": "psp-k8spsphostnetworkingports",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "hostNetwork": true,
      },
    },
  },
]
`;

exports[`transform_gatekeeper hostPID 1`] = `Array []`;

exports[`transform_gatekeeper hostPorts 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPHostNetworkingPorts",
    "metadata": Object {
      "name": "psp-k8spsphostnetworkingports",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "hostNetwork": true,
        "max": 9000,
        "min": 80,
      },
    },
  },
]
`;

exports[`transform_gatekeeper privileged 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPPrivilegedContainer",
    "metadata": Object {
      "name": "psp-k8spspprivilegedcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
]
`;

exports[`transform_gatekeeper readOnlyRootFilesystem 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPReadOnlyRootFilesystem",
    "metadata": Object {
      "name": "psp-k8spspreadonlyrootfilesystem",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
]
`;

exports[`transform_gatekeeper requiredDropCapabilities 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPCapabilities",
    "metadata": Object {
      "name": "psp-k8spspcapabilities",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedCapabilities": Array [],
        "requiredDropCapabilities": Array [
          "something",
        ],
      },
    },
  },
]
`;

exports[`transform_gatekeeper runAsGroup 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPPrivilegedContainer",
    "metadata": Object {
      "name": "psp-k8spspprivilegedcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPAllowedUsers",
    "metadata": Object {
      "name": "psp-k8spspallowedusers",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "runAsGroup": Object {
          "ranges": Array [
            Object {
              "max": 200,
              "min": 100,
            },
          ],
          "rule": "MustRunAs",
        },
      },
    },
  },
]
`;

exports[`transform_gatekeeper runAsUser 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPPrivilegedContainer",
    "metadata": Object {
      "name": "psp-k8spspprivilegedcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPAllowedUsers",
    "metadata": Object {
      "name": "psp-k8spspallowedusers",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "runAsUser": Object {
          "ranges": Array [
            Object {
              "max": 200,
              "min": 100,
            },
          ],
          "rule": "MustRunAs",
        },
      },
    },
  },
]
`;

exports[`transform_gatekeeper seLinux 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPSELinuxV2",
    "metadata": Object {
      "name": "psp-k8spspselinuxv2",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedSELinuxOptions": Array [
          Object {
            "level": "s0:c123,c456",
            "role": "object_r",
            "type": "svirt_sandbox_file_t",
            "user": "system_u",
          },
        ],
      },
    },
  },
]
`;

exports[`transform_gatekeeper seccomp 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPSeccomp",
    "metadata": Object {
      "name": "psp-k8spspseccomp",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "allowedProfiles": Array [
          "runtime/default",
          "docker/default",
        ],
      },
    },
  },
]
`;

exports[`transform_gatekeeper supplementalGroups 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPPrivilegedContainer",
    "metadata": Object {
      "name": "psp-k8spspprivilegedcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPAllowedUsers",
    "metadata": Object {
      "name": "psp-k8spspallowedusers",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "supplementalGroups": Object {
          "ranges": Array [
            Object {
              "max": 200,
              "min": 100,
            },
          ],
          "rule": "MustRunAs",
        },
      },
    },
  },
]
`;

exports[`transform_gatekeeper volumes 1`] = `
Array [
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPPrivilegedContainer",
    "metadata": Object {
      "name": "psp-k8spspprivilegedcontainer",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": null,
    },
  },
  Object {
    "apiVersion": "constraints.gatekeeper.sh/v1beta1",
    "kind": "K8sPSPVolumeTypes",
    "metadata": Object {
      "name": "psp-k8spspvolumetypes",
    },
    "spec": Object {
      "match": Object {
        "kinds": Array [
          Object {
            "apiGroups": Array [
              "",
            ],
            "kinds": Array [
              "Pod",
            ],
          },
        ],
      },
      "parameters": Object {
        "volumes": Array [
          "projected",
          "emptyDir",
        ],
      },
    },
  },
]
`;
